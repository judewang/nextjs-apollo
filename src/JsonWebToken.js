import _ from 'lodash';
import NodeRSA from 'node-rsa';
import cookie from 'cookie';
import { assertEnv } from 'thelper';
import jsonwebtoken from 'jsonwebtoken';
import { AuthenticationError } from 'apollo-server-express';
import generateSecret from './generateSecret';

const EXPIRED_IN = 14 * 24 * 60 * 60; // 7 days in seconds

export default class JsonWebToken {
  static generateKey() {
    return Buffer.from(new NodeRSA().generateKeyPair().exportKey('pkcs1-private-pem')).toString('base64');
  }

  constructor(env = {}, handlers = {}) {
    if (env.PRIVATE_KEY) {
      this.privateKey = Buffer.from(env.PRIVATE_KEY, 'base64');
      this.publicKey = new NodeRSA(this.privateKey).exportKey('pkcs8-public-pem');
      this.algorithm = { algorithm: 'RS256' };
    } else {
      this.privateKey = _.defaultTo(env.JWT_SECRET, env.HEROKU_APP_ID);
      this.publicKey = this.privateKey;
    }

    if (!this.privateKey) {
      throw new TypeError('PRIVATE_KEY or JWT_SECRET least one is required');
    }

    this.options = {
      expiresIn: _.defaultTo(env.JWT_EXPIRES_IN, EXPIRED_IN), // one hour in seconds
      issuer: _.defaultTo(env.JWT_ISSUER, 'jwt'),
      subject: _.defaultTo(env.JWT_ISSUER, 'jwt'),
      audience: _.defaultTo(env.JWT_AUDIENCE, 'everyone'),
    };

    _.assign(this, _.pick(_.defaults(handlers, {
      toValues: _.identity,
      toModel: _.identity,
      createCorrelation: _.identity,
      updateCorrelation: _.identity,
      fetchCorrelation: _.identity,
    }), [
      'toValues', 'toModel',
      'createCorrelation', 'updateCorrelation', 'fetchCorrelation',
    ]));

    return this;
  }

  sign(data, options) {
    const opt = _.defaults({}, options, this.options, this.algorithm);
    return jsonwebtoken.sign(data, this.privateKey, opt);
  }

  verify(token, options) {
    const opt = _.defaults({}, options, this.options, { algorithms: ['HS256', 'RS256'] });
    return jsonwebtoken.verify(token, this.publicKey, opt);
  }

  signWithCorrelation(correlation, user, ...args) {
    return this.sign({ correlation, user: this.toValues(user) }, ...args);
  }

  verifyWithCorrelation(token, ...args) {
    const { correlation, user, exp } = this.verify(token, ...args);
    const expiresOn = exp - (Date.now() / 1000);
    return { correlation, user: this.toModel(user), expiresOn };
  }

  async renewWithCorrelation({ id, secret }, auth, req) {
    const correlation = await this.fetchCorrelation(id, auth, req);
    if (!(correlation && correlation.secret === secret)) throw new Error('not match to secret');

    const renew = String(generateSecret());

    await this.updateCorrelation(id, renew, auth, req);
    return this.signWithCorrelation({ id, secret: renew }, auth.user);
  }

  async declareCorrelation(id, auth, req) {
    const correlation = id && await this.fetchCorrelation(id, auth, req);
    if (correlation) return correlation;

    return this.createCorrelation(auth, req);
  }

  async registerCorrelation(id, auth, req) {
    const correlation = await this.declareCorrelation(id, auth, req);

    if (correlation.isUnfamiliarCorrelation) return correlation;

    return this.updateCorrelation(id, String(generateSecret()), auth, req);
  }

  async convertTokenToUser(auth, req, res) {
    const { token, correlationId } = auth;
    if (!token) return null;

    const { options } = this;
    const { correlation, user, expiresOn } = this.verifyWithCorrelation(token);

    const isExpired = expiresOn < (options.expiresIn - (60 * 60));

    if (req) {
      if (!(correlationId && correlationId === correlation.id)) {
        throw new Error('not match to correlation id');
      }

      if (isExpired) {
        res.append('authorization', await this.renewWithCorrelation(correlation, { ...auth, user }, req));
      }

      return user;
    }

    if (isExpired) throw new Error('token is not safe');
    return user;
  }

  contextParser() {
    return async (context) => {
      const { req, res, connection } = context;

      const auth = {
        signIn: async (user) => {
          auth.user = user;

          const correlation = await this.registerCorrelation(auth.correlationId, auth, req);

          if (auth.correlationId !== correlation.id) {
            res.append('Set-Cookie', cookie.serialize(
              'x-correlation-id',
              correlation.id,
              { httpOnly: true, maxAge: 365 * 24 * 60 * 60 },
            ));
          }

          if (!correlation.isUnfamiliarCorrelation) {
            const { id, secret } = correlation;
            const token = this.signWithCorrelation({ id, secret }, auth.user);
            res.append('authorization', token);
            return { correlation, token };
          }

          return { correlation };
        },
      };

      try {
        const { authorization } = connection ? connection.context : req.headers;
        const [, token] = /^Bearer (.+)$/.exec(authorization);
        auth.token = token;
      } catch (e) {
        // empty
      }

      try {
        auth.correlationId = cookie.parse(req.headers.cookie)['x-correlation-id'];
      } catch (e) {
        // empty
      }

      try {
        const user = await this.convertTokenToUser(auth, req, res);
        auth.user = user;
        return { ...context, auth };
      } catch (e) {
        assertEnv(() => {
          if (!res) throw new AuthenticationError(e.message);
          res.append('X-Content-Extend', e.message);
        });
        throw new AuthenticationError('must authenticate');
      }
    };
  }
}
