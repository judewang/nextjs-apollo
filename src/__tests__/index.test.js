import JsonWebToken from '../JsonWebToken';

describe('jwt', () => {
  it('is required', () => {
    expect(() => new JsonWebToken()).toThrowError(new TypeError('PRIVATE_KEY or JWT_SECRET least one is required'));
  });

  it('use JWT_SECRET to sign & verify', async () => {
    const jwt = new JsonWebToken({ JWT_SECRET: 'XYZ' });
    expect(jwt).toMatchSnapshot();
    const { data } = jwt.verify(jwt.sign({ id: '10' }));
    expect(data.id).toBe('10');
  });

  it('use PRIVATE_KEY to sign & verify', async () => {
    const PRIVATE_KEY = 'LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFb3dJQkFBS0NBUUVBc0RWMnB6azIydTRGTGtyTW9TSjZpQnEwQmxSS1dVMSsxc0VrSG5ycWNsNUhxeEp4CmR4elBkS1dBRTV6TVJqeUlRTTlKSWcxTEllVGkzNE1paGRrbjNHaHIydXVtOWk0eXJ3VVZpRVBDbWNxS0JsekoKYjB2a3YrRUZYcHc4VVhJblFMZzlyWkdkZlFQTUtUUlVJYVNFYmxMcE5tTjJ6MGJ5L1M4aXBIN3QrelRablRkZgoyQVBQN3U5NGpEaVA2QWc1ZHlGaUpVejRlQTNEZUlUbkZwMjE4N1VTZkJMUTFmSVB2NXNtM2RJSkZWUFhwNHFKClM0R3BiRDhIamxKUEdXdFhwRHRDeFlvTkFHbGRiY3VEcTY0VGJWTEFBTHlZcXFMVkV6bmpOdUdwc0N0S0dva2oKSjJoRlZDbXQwQ0JGSWxJQnRiZVBSaC94Q05kdzRLMFJzN0ozc3dJREFRQUJBb0lCQURwZEN2YnFhWFBzUUlVTAp1MmxuclNRbGpCbElrL091U21QQlIwQmRRQWd2bUJkZW5YeHhhM2xjVSthckdiSHQzTWR5b091SmNhaEJVcS9yCmxHUzIvNC91ZWtMaHFNU0Rtc0FEdEJVVk5JQ1I2Rk9oS2dzV0hTYzgxZlYzVkRjNTB5eUptY3hDbmVKSmZyYnIKSVM2ZmVrbThab2hnTzNyK0tmK1lQQ1kxOEljeWRqb2FwVmJ0MTlpdmUxb1dIblIyWkNVQk05WlF2RklzZTlRZgpncFk3Ly96QjlxRUlqMEMzMGsxT0tMd1k3WnAxUWlhMUl0TzYwYzdrVzFNZlVGL3VMZ3dldld4STZ1SnNWaFNYCk5sd3hsTzgvTkhHazE2bEMyZWtEM2pWcWpjSHdDcCt2VjFRdkFaUElvOUVEMGozVFlhZFRiWDcwQjNGRS9uU0YKL2R3am4xa0NnWUVBMzhwTG1DWTFxdGZ6MGdCUWJrNnA1a0ZsaktGOGd3QlhPWkM0TnJCU0F5bFgyK3dUWFdTTQpGVTRMNHJlL0EzMTlHVkYvT2RlQitDOERXVFVtQ1Y5SHZYaDNPQzA5VGlEekFoYWdPOTlaMWF0cFc4c2xaa2ZICk9TU01nNmxjUFd3em5WRW5FNUJZRGdMKzJFamtlR1pJZjR2NWpiem44N044cGRWM1d0L1pRTThDZ1lFQXlaSC8KNndKVUlmaFcvWUVyVCtHbTl5Vk5KNUZHa0FRRmxBZGo0bDhTT2ZVMTBZbzlqVUZvOWpoVkhNRng1MDJUeGZGNgphL2lodzF0NE04OFdnZlFKekdoTjBWQmVIWVhnZDlpWk9DckZoQUJNR3ltZzhrYWVTd1RwRkt4SUFDeEQycGgyCjNVWHJxOHFkRHVWRVVzaG1yc1R1Vm11dWlGRERDZHg5bzBYQmE5MENnWUVBczhWbjhITlJUVXBvL050dndRekUKQXNtcDc4eVJueEtZL3c4NFRBWmIrMW9zeitmV3JvRmszb0J4bjlDSFN5QjFhKzlCeW85S1FycFJjRW4zb2pPKwpCNEJNRlA2U2JyNmRONkpCV1pkVUxRYWpmaSswZlhOUG9LR3hsdHBuMUplazVzMFlIT2twTXloSTNDbkNEczNyCmt6bzRZQm5PSXNDZWN3RUYvaVF2KzVVQ2dZQnI3YXZkRnRNYm1WQml6bzMxYUxxdzdKMGlTWXhCd3BXR0RHcjAKSTdtcnBOdGJSaXpzL0pBSi9tRzhVdXY5VUxHR05HVTdlTmFNdHVabXYwU3E4cE5vaXBUWU5iTWZGYmI2Rk5UOQptUkZITzYyRWdjYTIwdzhnMWhGRFMzZ0Y5NjJoVnkrV0RVRkNLK3ZxdE9IZmZZUlZvZ2FINllKN1pOcjU1Y0N6CjhyNjlWUUtCZ0gyb2FMMURHNnlMc1hzYjVsa2Jvb3BtMUVNdjhQMWQ1cy9XOVlNSy9oZWRtSTY1TmN1a0hVNkQKdGpZa25xaWliL1ZYNTAyWmIvTkFxY1ZEYXRMMXJqbVFnalUvVnBSYlNEQ2NRek5OTmhpN1phWldHZXovbWE4QQpXcEhVWXdZSzlBZSsvMnRTQWFZaHJ4M0s2VWM0clFwK0cwS3c1blRsZlV3V2srTVltSEpoCi0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0t';

    const jwt = new JsonWebToken({ PRIVATE_KEY });
    expect(jwt).toMatchSnapshot();
    const { data } = jwt.verify(jwt.sign({ id: '10', acr: 'NONO' }));
    expect(data.id).toBe('10');
  });

  it('setup JWT_ISSUER, JWT_EXPIRES_IN and JWT_AUDIENCE', () => {
    const jwt = new JsonWebToken({
      JWT_SECRET: 'XYZ',
      JWT_ISSUER: 'tester',
      JWT_EXPIRES_IN: 24 * 60 * 60,
      JWT_AUDIENCE: 'for_tester',
    });
    expect(jwt).toMatchSnapshot();
  });

  it('generate key', () => {
    const key = JsonWebToken.generateKey();

    const jwt = new JsonWebToken({ PRIVATE_KEY: key });
    const { data } = jwt.verify(jwt.sign({ id: '10' }));
    expect(data.id).toBe('10');
  });
});
