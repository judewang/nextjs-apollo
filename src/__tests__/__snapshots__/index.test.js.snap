// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jwt setup JWT_ISSUER, JWT_EXPIRES_IN and JWT_AUDIENCE 1`] = `
JsonWebToken {
  "createCorrelation": [Function],
  "fetchCorrelation": [Function],
  "options": Object {
    "audience": "for_tester",
    "expiresIn": 86400,
    "issuer": "tester",
    "subject": "tester",
  },
  "privateKey": "XYZ",
  "publicKey": "XYZ",
  "toModel": [Function],
  "toValues": [Function],
  "updateCorrelation": [Function],
}
`;

exports[`jwt use JWT_SECRET to sign & verify 1`] = `
JsonWebToken {
  "createCorrelation": [Function],
  "fetchCorrelation": [Function],
  "options": Object {
    "audience": "everyone",
    "expiresIn": 1209600,
    "issuer": "jwt",
    "subject": "jwt",
  },
  "privateKey": "XYZ",
  "publicKey": "XYZ",
  "toModel": [Function],
  "toValues": [Function],
  "updateCorrelation": [Function],
}
`;

exports[`jwt use PRIVATE_KEY to sign & verify 1`] = `
JsonWebToken {
  "algorithm": Object {
    "algorithm": "RS256",
  },
  "createCorrelation": [Function],
  "fetchCorrelation": [Function],
  "options": Object {
    "audience": "everyone",
    "expiresIn": 1209600,
    "issuer": "jwt",
    "subject": "jwt",
  },
  "privateKey": Object {
    "data": Array [
      45,
      45,
      45,
      45,
      45,
      66,
      69,
      71,
      73,
      78,
      32,
      82,
      83,
      65,
      32,
      80,
      82,
      73,
      86,
      65,
      84,
      69,
      32,
      75,
      69,
      89,
      45,
      45,
      45,
      45,
      45,
      10,
      77,
      73,
      73,
      69,
      111,
      119,
      73,
      66,
      65,
      65,
      75,
      67,
      65,
      81,
      69,
      65,
      115,
      68,
      86,
      50,
      112,
      122,
      107,
      50,
      50,
      117,
      52,
      70,
      76,
      107,
      114,
      77,
      111,
      83,
      74,
      54,
      105,
      66,
      113,
      48,
      66,
      108,
      82,
      75,
      87,
      85,
      49,
      43,
      49,
      115,
      69,
      107,
      72,
      110,
      114,
      113,
      99,
      108,
      53,
      72,
      113,
      120,
      74,
      120,
      10,
      100,
      120,
      122,
      80,
      100,
      75,
      87,
      65,
      69,
      53,
      122,
      77,
      82,
      106,
      121,
      73,
      81,
      77,
      57,
      74,
      73,
      103,
      49,
      76,
      73,
      101,
      84,
      105,
      51,
      52,
      77,
      105,
      104,
      100,
      107,
      110,
      51,
      71,
      104,
      114,
      50,
      117,
      117,
      109,
      57,
      105,
      52,
      121,
      114,
      119,
      85,
      86,
      105,
      69,
      80,
      67,
      109,
      99,
      113,
      75,
      66,
      108,
      122,
      74,
      10,
      98,
      48,
      118,
      107,
      118,
      43,
      69,
      70,
      88,
      112,
      119,
      56,
      85,
      88,
      73,
      110,
      81,
      76,
      103,
      57,
      114,
      90,
      71,
      100,
      102,
      81,
      80,
      77,
      75,
      84,
      82,
      85,
      73,
      97,
      83,
      69,
      98,
      108,
      76,
      112,
      78,
      109,
      78,
      50,
      122,
      48,
      98,
      121,
      47,
      83,
      56,
      105,
      112,
      72,
      55,
      116,
      43,
      122,
      84,
      90,
      110,
      84,
      100,
      102,
      10,
      50,
      65,
      80,
      80,
      55,
      117,
      57,
      52,
      106,
      68,
      105,
      80,
      54,
      65,
      103,
      53,
      100,
      121,
      70,
      105,
      74,
      85,
      122,
      52,
      101,
      65,
      51,
      68,
      101,
      73,
      84,
      110,
      70,
      112,
      50,
      49,
      56,
      55,
      85,
      83,
      102,
      66,
      76,
      81,
      49,
      102,
      73,
      80,
      118,
      53,
      115,
      109,
      51,
      100,
      73,
      74,
      70,
      86,
      80,
      88,
      112,
      52,
      113,
      74,
      10,
      83,
      52,
      71,
      112,
      98,
      68,
      56,
      72,
      106,
      108,
      74,
      80,
      71,
      87,
      116,
      88,
      112,
      68,
      116,
      67,
      120,
      89,
      111,
      78,
      65,
      71,
      108,
      100,
      98,
      99,
      117,
      68,
      113,
      54,
      52,
      84,
      98,
      86,
      76,
      65,
      65,
      76,
      121,
      89,
      113,
      113,
      76,
      86,
      69,
      122,
      110,
      106,
      78,
      117,
      71,
      112,
      115,
      67,
      116,
      75,
      71,
      111,
      107,
      106,
      10,
      74,
      50,
      104,
      70,
      86,
      67,
      109,
      116,
      48,
      67,
      66,
      70,
      73,
      108,
      73,
      66,
      116,
      98,
      101,
      80,
      82,
      104,
      47,
      120,
      67,
      78,
      100,
      119,
      52,
      75,
      48,
      82,
      115,
      55,
      74,
      51,
      115,
      119,
      73,
      68,
      65,
      81,
      65,
      66,
      65,
      111,
      73,
      66,
      65,
      68,
      112,
      100,
      67,
      118,
      98,
      113,
      97,
      88,
      80,
      115,
      81,
      73,
      85,
      76,
      10,
      117,
      50,
      108,
      110,
      114,
      83,
      81,
      108,
      106,
      66,
      108,
      73,
      107,
      47,
      79,
      117,
      83,
      109,
      80,
      66,
      82,
      48,
      66,
      100,
      81,
      65,
      103,
      118,
      109,
      66,
      100,
      101,
      110,
      88,
      120,
      120,
      97,
      51,
      108,
      99,
      85,
      43,
      97,
      114,
      71,
      98,
      72,
      116,
      51,
      77,
      100,
      121,
      111,
      79,
      117,
      74,
      99,
      97,
      104,
      66,
      85,
      113,
      47,
      114,
      10,
      108,
      71,
      83,
      50,
      47,
      52,
      47,
      117,
      101,
      107,
      76,
      104,
      113,
      77,
      83,
      68,
      109,
      115,
      65,
      68,
      116,
      66,
      85,
      86,
      78,
      73,
      67,
      82,
      54,
      70,
      79,
      104,
      75,
      103,
      115,
      87,
      72,
      83,
      99,
      56,
      49,
      102,
      86,
      51,
      86,
      68,
      99,
      53,
      48,
      121,
      121,
      74,
      109,
      99,
      120,
      67,
      110,
      101,
      74,
      74,
      102,
      114,
      98,
      114,
      10,
      73,
      83,
      54,
      102,
      101,
      107,
      109,
      56,
      90,
      111,
      104,
      103,
      79,
      51,
      114,
      43,
      75,
      102,
      43,
      89,
      80,
      67,
      89,
      49,
      56,
      73,
      99,
      121,
      100,
      106,
      111,
      97,
      112,
      86,
      98,
      116,
      49,
      57,
      105,
      118,
      101,
      49,
      111,
      87,
      72,
      110,
      82,
      50,
      90,
      67,
      85,
      66,
      77,
      57,
      90,
      81,
      118,
      70,
      73,
      115,
      101,
      57,
      81,
      102,
      10,
      103,
      112,
      89,
      55,
      47,
      47,
      122,
      66,
      57,
      113,
      69,
      73,
      106,
      48,
      67,
      51,
      48,
      107,
      49,
      79,
      75,
      76,
      119,
      89,
      55,
      90,
      112,
      49,
      81,
      105,
      97,
      49,
      73,
      116,
      79,
      54,
      48,
      99,
      55,
      107,
      87,
      49,
      77,
      102,
      85,
      70,
      47,
      117,
      76,
      103,
      119,
      101,
      118,
      87,
      120,
      73,
      54,
      117,
      74,
      115,
      86,
      104,
      83,
      88,
      10,
      78,
      108,
      119,
      120,
      108,
      79,
      56,
      47,
      78,
      72,
      71,
      107,
      49,
      54,
      108,
      67,
      50,
      101,
      107,
      68,
      51,
      106,
      86,
      113,
      106,
      99,
      72,
      119,
      67,
      112,
      43,
      118,
      86,
      49,
      81,
      118,
      65,
      90,
      80,
      73,
      111,
      57,
      69,
      68,
      48,
      106,
      51,
      84,
      89,
      97,
      100,
      84,
      98,
      88,
      55,
      48,
      66,
      51,
      70,
      69,
      47,
      110,
      83,
      70,
      10,
      47,
      100,
      119,
      106,
      110,
      49,
      107,
      67,
      103,
      89,
      69,
      65,
      51,
      56,
      112,
      76,
      109,
      67,
      89,
      49,
      113,
      116,
      102,
      122,
      48,
      103,
      66,
      81,
      98,
      107,
      54,
      112,
      53,
      107,
      70,
      108,
      106,
      75,
      70,
      56,
      103,
      119,
      66,
      88,
      79,
      90,
      67,
      52,
      78,
      114,
      66,
      83,
      65,
      121,
      108,
      88,
      50,
      43,
      119,
      84,
      88,
      87,
      83,
      77,
      10,
      70,
      85,
      52,
      76,
      52,
      114,
      101,
      47,
      65,
      51,
      49,
      57,
      71,
      86,
      70,
      47,
      79,
      100,
      101,
      66,
      43,
      67,
      56,
      68,
      87,
      84,
      85,
      109,
      67,
      86,
      57,
      72,
      118,
      88,
      104,
      51,
      79,
      67,
      48,
      57,
      84,
      105,
      68,
      122,
      65,
      104,
      97,
      103,
      79,
      57,
      57,
      90,
      49,
      97,
      116,
      112,
      87,
      56,
      115,
      108,
      90,
      107,
      102,
      72,
      10,
      79,
      83,
      83,
      77,
      103,
      54,
      108,
      99,
      80,
      87,
      119,
      122,
      110,
      86,
      69,
      110,
      69,
      53,
      66,
      89,
      68,
      103,
      76,
      43,
      50,
      69,
      106,
      107,
      101,
      71,
      90,
      73,
      102,
      52,
      118,
      53,
      106,
      98,
      122,
      110,
      56,
      55,
      78,
      56,
      112,
      100,
      86,
      51,
      87,
      116,
      47,
      90,
      81,
      77,
      56,
      67,
      103,
      89,
      69,
      65,
      121,
      90,
      72,
      47,
      10,
      54,
      119,
      74,
      85,
      73,
      102,
      104,
      87,
      47,
      89,
      69,
      114,
      84,
      43,
      71,
      109,
      57,
      121,
      86,
      78,
      74,
      53,
      70,
      71,
      107,
      65,
      81,
      70,
      108,
      65,
      100,
      106,
      52,
      108,
      56,
      83,
      79,
      102,
      85,
      49,
      48,
      89,
      111,
      57,
      106,
      85,
      70,
      111,
      57,
      106,
      104,
      86,
      72,
      77,
      70,
      120,
      53,
      48,
      50,
      84,
      120,
      102,
      70,
      54,
      10,
      97,
      47,
      105,
      104,
      119,
      49,
      116,
      52,
      77,
      56,
      56,
      87,
      103,
      102,
      81,
      74,
      122,
      71,
      104,
      78,
      48,
      86,
      66,
      101,
      72,
      89,
      88,
      103,
      100,
      57,
      105,
      90,
      79,
      67,
      114,
      70,
      104,
      65,
      66,
      77,
      71,
      121,
      109,
      103,
      56,
      107,
      97,
      101,
      83,
      119,
      84,
      112,
      70,
      75,
      120,
      73,
      65,
      67,
      120,
      68,
      50,
      112,
      104,
      50,
      10,
      51,
      85,
      88,
      114,
      113,
      56,
      113,
      100,
      68,
      117,
      86,
      69,
      85,
      115,
      104,
      109,
      114,
      115,
      84,
      117,
      86,
      109,
      117,
      117,
      105,
      70,
      68,
      68,
      67,
      100,
      120,
      57,
      111,
      48,
      88,
      66,
      97,
      57,
      48,
      67,
      103,
      89,
      69,
      65,
      115,
      56,
      86,
      110,
      56,
      72,
      78,
      82,
      84,
      85,
      112,
      111,
      47,
      78,
      116,
      118,
      119,
      81,
      122,
      69,
      10,
      65,
      115,
      109,
      112,
      55,
      56,
      121,
      82,
      110,
      120,
      75,
      89,
      47,
      119,
      56,
      52,
      84,
      65,
      90,
      98,
      43,
      49,
      111,
      115,
      122,
      43,
      102,
      87,
      114,
      111,
      70,
      107,
      51,
      111,
      66,
      120,
      110,
      57,
      67,
      72,
      83,
      121,
      66,
      49,
      97,
      43,
      57,
      66,
      121,
      111,
      57,
      75,
      81,
      114,
      112,
      82,
      99,
      69,
      110,
      51,
      111,
      106,
      79,
      43,
      10,
      66,
      52,
      66,
      77,
      70,
      80,
      54,
      83,
      98,
      114,
      54,
      100,
      78,
      54,
      74,
      66,
      87,
      90,
      100,
      85,
      76,
      81,
      97,
      106,
      102,
      105,
      43,
      48,
      102,
      88,
      78,
      80,
      111,
      75,
      71,
      120,
      108,
      116,
      112,
      110,
      49,
      74,
      101,
      107,
      53,
      115,
      48,
      89,
      72,
      79,
      107,
      112,
      77,
      121,
      104,
      73,
      51,
      67,
      110,
      67,
      68,
      115,
      51,
      114,
      10,
      107,
      122,
      111,
      52,
      89,
      66,
      110,
      79,
      73,
      115,
      67,
      101,
      99,
      119,
      69,
      70,
      47,
      105,
      81,
      118,
      43,
      53,
      85,
      67,
      103,
      89,
      66,
      114,
      55,
      97,
      118,
      100,
      70,
      116,
      77,
      98,
      109,
      86,
      66,
      105,
      122,
      111,
      51,
      49,
      97,
      76,
      113,
      119,
      55,
      74,
      48,
      105,
      83,
      89,
      120,
      66,
      119,
      112,
      87,
      71,
      68,
      71,
      114,
      48,
      10,
      73,
      55,
      109,
      114,
      112,
      78,
      116,
      98,
      82,
      105,
      122,
      115,
      47,
      74,
      65,
      74,
      47,
      109,
      71,
      56,
      85,
      117,
      118,
      57,
      85,
      76,
      71,
      71,
      78,
      71,
      85,
      55,
      101,
      78,
      97,
      77,
      116,
      117,
      90,
      109,
      118,
      48,
      83,
      113,
      56,
      112,
      78,
      111,
      105,
      112,
      84,
      89,
      78,
      98,
      77,
      102,
      70,
      98,
      98,
      54,
      70,
      78,
      84,
      57,
      10,
      109,
      82,
      70,
      72,
      79,
      54,
      50,
      69,
      103,
      99,
      97,
      50,
      48,
      119,
      56,
      103,
      49,
      104,
      70,
      68,
      83,
      51,
      103,
      70,
      57,
      54,
      50,
      104,
      86,
      121,
      43,
      87,
      68,
      85,
      70,
      67,
      75,
      43,
      118,
      113,
      116,
      79,
      72,
      102,
      102,
      89,
      82,
      86,
      111,
      103,
      97,
      72,
      54,
      89,
      74,
      55,
      90,
      78,
      114,
      53,
      53,
      99,
      67,
      122,
      10,
      56,
      114,
      54,
      57,
      86,
      81,
      75,
      66,
      103,
      72,
      50,
      111,
      97,
      76,
      49,
      68,
      71,
      54,
      121,
      76,
      115,
      88,
      115,
      98,
      53,
      108,
      107,
      98,
      111,
      111,
      112,
      109,
      49,
      69,
      77,
      118,
      56,
      80,
      49,
      100,
      53,
      115,
      47,
      87,
      57,
      89,
      77,
      75,
      47,
      104,
      101,
      100,
      109,
      73,
      54,
      53,
      78,
      99,
      117,
      107,
      72,
      85,
      54,
      68,
      10,
      116,
      106,
      89,
      107,
      110,
      113,
      105,
      105,
      98,
      47,
      86,
      88,
      53,
      48,
      50,
      90,
      98,
      47,
      78,
      65,
      113,
      99,
      86,
      68,
      97,
      116,
      76,
      49,
      114,
      106,
      109,
      81,
      103,
      106,
      85,
      47,
      86,
      112,
      82,
      98,
      83,
      68,
      67,
      99,
      81,
      122,
      78,
      78,
      78,
      104,
      105,
      55,
      90,
      97,
      90,
      87,
      71,
      101,
      122,
      47,
      109,
      97,
      56,
      65,
      10,
      87,
      112,
      72,
      85,
      89,
      119,
      89,
      75,
      57,
      65,
      101,
      43,
      47,
      50,
      116,
      83,
      65,
      97,
      89,
      104,
      114,
      120,
      51,
      75,
      54,
      85,
      99,
      52,
      114,
      81,
      112,
      43,
      71,
      48,
      75,
      119,
      53,
      110,
      84,
      108,
      102,
      85,
      119,
      87,
      107,
      43,
      77,
      89,
      109,
      72,
      74,
      104,
      10,
      45,
      45,
      45,
      45,
      45,
      69,
      78,
      68,
      32,
      82,
      83,
      65,
      32,
      80,
      82,
      73,
      86,
      65,
      84,
      69,
      32,
      75,
      69,
      89,
      45,
      45,
      45,
      45,
      45,
    ],
    "type": "Buffer",
  },
  "publicKey": "-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAsDV2pzk22u4FLkrMoSJ6
iBq0BlRKWU1+1sEkHnrqcl5HqxJxdxzPdKWAE5zMRjyIQM9JIg1LIeTi34Mihdkn
3Ghr2uum9i4yrwUViEPCmcqKBlzJb0vkv+EFXpw8UXInQLg9rZGdfQPMKTRUIaSE
blLpNmN2z0by/S8ipH7t+zTZnTdf2APP7u94jDiP6Ag5dyFiJUz4eA3DeITnFp21
87USfBLQ1fIPv5sm3dIJFVPXp4qJS4GpbD8HjlJPGWtXpDtCxYoNAGldbcuDq64T
bVLAALyYqqLVEznjNuGpsCtKGokjJ2hFVCmt0CBFIlIBtbePRh/xCNdw4K0Rs7J3
swIDAQAB
-----END PUBLIC KEY-----",
  "toModel": [Function],
  "toValues": [Function],
  "updateCorrelation": [Function],
}
`;
